// @dcloudio/uni-app-uts/dist/plugins/js/plugin.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createUniAppJsEnginePlugin = exports.initUniAppJsEngineDom1CssPlugin = void 0;
const path_1 = __importDefault(require("path"));
const node_crypto_1 = require("node:crypto");
const fs_extra_1 = __importDefault(require("fs-extra"));
const uni_cli_shared_1 = require("@dcloudio/uni-cli-shared");
const utils_1 = require("../utils");
const css_1 = require("./css");
const js_1 = require("./js");
const sourceMap_1 = require("./sourceMap");
const HARMONY_DOM2_ESBUILD_TRANSPILE_CACHE_MAX = 2048;
const HARMONY_DOM2_ESBUILD_TRANSPILE_CACHE_FLAG = '__uni_harmony_dom2_esbuild_transpile_cache__';
// 鸿蒙 dom2 增量编译时，Rollup 每次 generate 都会让 vite:esbuild-transpile
// 对所有 chunk 重新执行 renderChunk。多数未变更 chunk 的 code、fileName、format 都相同，
// 这部分 esbuild 转译是重复开销，性能报告里它也是 generate 阶段最大的单点耗时。
//
// 这里不替换 Vite 的实现，只在鸿蒙 dom2 下包装原 renderChunk：
// 1. 命中缓存时直接复用原 renderChunk 的返回值，跳过重复 esbuild transform；
// 2. 未命中时仍调用 Vite 原逻辑，保证功能和 sourcemap 处理不变；
// 3. 缓存只放在当前 Vite 配置实例内，不跨进程、不落盘，避免引入额外失效问题。
function initHarmonyDom2EsbuildTranspileCache(config) {
    const plugin = config.plugins.find((p) => p.name === 'vite:esbuild-transpile');
    if (!plugin ||
        !plugin.renderChunk ||
        plugin[HARMONY_DOM2_ESBUILD_TRANSPILE_CACHE_FLAG]) {
        return;
    }
    const rawRenderChunk = plugin.renderChunk;
    const rawHandler = typeof rawRenderChunk === 'function'
        ? rawRenderChunk
        : rawRenderChunk.handler;
    if (typeof rawHandler !== 'function') {
        return;
    }
    const cache = new Map();
    const renderChunk = function (code, chunk, opts) {
        // plugin-legacy 会通过该标记跳过 esbuild，保持原逻辑直接透传。
        if (opts?.__vite_skip_esbuild__) {
            return rawHandler.call(this, code, chunk, opts);
        }
        const cacheKey = createEsbuildTranspileCacheKey(code, chunk, opts);
        if (cache.has(cacheKey)) {
            const cached = cache.get(cacheKey);
            // Map 充当简单 LRU：命中后挪到末尾，避免热 chunk 被淘汰。
            cache.delete(cacheKey);
            cache.set(cacheKey, cached);
            return cached;
        }
        const result = Promise.resolve(rawHandler.call(this, code, chunk, opts))
            .then((res) => {
            // 缓存原 hook 的完整结果，而不是只缓存 code，避免遗漏 map 等字段。
            cache.set(cacheKey, res);
            return res;
        })
            .catch((err) => {
            // 转译失败不能污染缓存，避免后续修复代码后仍复用错误结果。
            cache.delete(cacheKey);
            throw err;
        });
        cache.set(cacheKey, result);
        trimEsbuildTranspileCache(cache);
        return result;
    };
    plugin.renderChunk =
        typeof rawRenderChunk === 'function'
            ? renderChunk
            : { ...rawRenderChunk, handler: renderChunk };
    plugin[HARMONY_DOM2_ESBUILD_TRANSPILE_CACHE_FLAG] = true;
}
function createEsbuildTranspileCacheKey(code, chunk, opts) {
    // renderChunk 的输出主要由 chunk 内容和输出参数决定：
    // - code hash/length：判断 chunk 内容是否真的没变；
    // - fileName：Vite 会作为 esbuild sourcefile，影响 sourcemap；
    // - format/sourcemap：影响 esbuild 输出格式和 map 结果。
    // 只把 hash 放入 key，避免把大段 chunk code 长期挂在 Map key 上。
    const hash = (0, node_crypto_1.createHash)('sha1').update(code).digest('hex');
    return [
        opts?.format,
        opts?.sourcemap,
        chunk?.fileName,
        code.length,
        hash,
    ].join('|');
}
function trimEsbuildTranspileCache(cache) {
    // watch 场景中 chunk 会随编辑持续产生新内容，限制上限避免缓存无限增长。
    while (cache.size > HARMONY_DOM2_ESBUILD_TRANSPILE_CACHE_MAX) {
        const firstKey = cache.keys().next().value;
        if (!firstKey) {
            break;
        }
        cache.delete(firstKey);
    }
}
function initUniAppJsEngineDom1CssPlugin(config) {
    (0, uni_cli_shared_1.injectCssPlugin)(config, process.env.UNI_COMPILE_TARGET === 'uni_modules'
        ? {
            createUrlReplacer: uni_cli_shared_1.createEncryptCssUrlReplacer,
        }
        : {});
    (0, uni_cli_shared_1.injectCssPostPlugin)(config, (0, css_1.uniAppCssPlugin)(config));
}
exports.initUniAppJsEngineDom1CssPlugin = initUniAppJsEngineDom1CssPlugin;
function createUniAppJsEnginePlugin(platform) {
    return function uniAppJsEnginePlugin() {
        const inputDir = (0, uni_cli_shared_1.normalizePath)(process.env.UNI_INPUT_DIR);
        const outputDir = process.env.UNI_OUTPUT_DIR;
        const uvueOutputDir = (0, uni_cli_shared_1.uvueOutDir)(platform);
        const tscOutputDir = (0, uni_cli_shared_1.tscOutDir)(platform);
        // 开始编译时，清空输出目录
        function emptyOutDir() {
            // ext-api 编译时，需要同时编译多个平台，并保留多个平台的输出目录
            if (process.env.UNI_COMPILE_TARGET === 'ext-api') {
                return;
            }
            if (fs_extra_1.default.existsSync(outputDir)) {
                (0, uni_cli_shared_1.emptyDir)(outputDir);
            }
        }
        emptyOutDir();
        function emptyUVueDir() {
            if (fs_extra_1.default.existsSync(uvueOutputDir)) {
                (0, uni_cli_shared_1.emptyDir)(uvueOutputDir);
            }
        }
        emptyUVueDir();
        function emptyTscDir() {
            if (fs_extra_1.default.existsSync(tscOutputDir)) {
                (0, uni_cli_shared_1.emptyDir)(tscOutputDir);
            }
        }
        emptyTscDir();
        if (process.env.UNI_UTS_PLATFORM === 'app-harmony' &&
            (0, uni_cli_shared_1.isNormalCompileTarget)()) {
            // 拆分模式
            process.env.UNI_APP_OUTPUT_FORMAT = 'esm';
            // 动态导入
            process.env.UNI_APP_DYNAMIC_IMPORT = 'true';
        }
        const isESM = process.env.UNI_APP_OUTPUT_FORMAT === 'esm';
        const paths = isESM
            ? {
                // vue: '@dcloudio/uni-app-x-runtime',
                '@vue/shared': '@dcloudio/uni-app-x-runtime',
            }
            : {};
        const isDom2 = process.env.UNI_APP_X_DOM2 === 'true';
        const isAndroid = platform === 'app-android';
        const isIOS = platform === 'app-ios';
        const isHarmony = platform === 'app-harmony';
        // 只缓存 sourcemap 内容 hash，不缓存完整内容，避免用空间换时间。
        const sourceMapHashCache = new Map();
        // 仅限定鸿蒙 DOM2 开发模式，避免影响其它平台和生产构建。
        const enableSourceMapIncremental = process.env.NODE_ENV === 'development' && isDom2 && isHarmony;
        const globals = {
            vue: 'Vue',
            '@vue/shared': 'uni.VueShared',
        };
        if (isDom2 && (isAndroid || isIOS)) {
            globals[utils_1.SHARED_DATA_LIB_NAME] = '__uniSharedDataLib';
        }
        return {
            name: 'uni:app-uts',
            apply: 'build',
            uni: (0, utils_1.createUniOptions)(platform),
            config(config) {
                const sourcemap = (0, uni_cli_shared_1.withSourcemap)(config);
                return {
                    base: '/', // 强制 base
                    build: {
                        sourcemap,
                        emptyOutDir: false,
                        assetsInlineLimit: 0,
                        target: isIOS
                            ? [
                                'ios12',
                                'es2020',
                                'edge88',
                                'firefox78',
                                'chrome87',
                                'safari14',
                            ]
                            : isHarmony
                                ? ['es2022']
                                : undefined,
                        rollupOptions: {
                            input: (0, uni_cli_shared_1.resolveMainPathOnce)(inputDir),
                            // import "libentry.so"
                            external: ['vue', '@vue/shared', /.*\.so$/],
                            output: {
                                name: 'AppService',
                                banner: ``,
                                format: isESM ? 'esm' : 'iife',
                                entryFileNames: uni_cli_shared_1.APP_SERVICE_FILENAME,
                                globals,
                                paths,
                                manualChunks(id) {
                                    if (isESM) {
                                        const chunkName = (0, uni_cli_shared_1.normalizePath)(id.split('?')[0]);
                                        if (chunkName.startsWith('\0plugin-vue:')) {
                                            return 'plugin-vue-' + chunkName.split(':')[1];
                                        }
                                        if (chunkName.includes('/@dcloudio/uni-cloud/')) {
                                            return '@dcloudio/uni-cloud';
                                        }
                                        if (chunkName.startsWith(inputDir)) {
                                            return (0, uni_cli_shared_1.removeExt)((0, uni_cli_shared_1.normalizePath)(path_1.default.relative(inputDir, chunkName)));
                                        }
                                    }
                                },
                                inlineDynamicImports: false,
                                chunkFileNames: isESM ? 'assets/[name].js' : undefined,
                                sourcemapPathTransform: (relativeSourcePath, sourcemapPath) => {
                                    return (0, uni_cli_shared_1.normalizePath)(path_1.default.relative(process.env.UNI_INPUT_DIR, path_1.default.resolve(path_1.default.dirname(sourcemapPath), relativeSourcePath)));
                                },
                            },
                            treeshake: {
                                moduleSideEffects: (id, external) => {
                                    if (id === '@vue/shared') {
                                        return false;
                                    }
                                    return true;
                                },
                            },
                        },
                    },
                };
            },
            configResolved(config) {
                (0, utils_1.configResolved)(config);
                if (!isDom2) {
                    initUniAppJsEngineDom1CssPlugin(config);
                }
                (0, uni_cli_shared_1.insertBeforePlugin)((0, js_1.uniAppJsPlugin)(config), 'uni:app-main', config);
                // 如果开启了 vapor 模式，则禁用 vue 的 devtools，让 @vitejs/plugin-vue 不管是开发还是发行，均生成发行代码
                // 理论上非 vapor 也应该禁用，但为了不引发其他问题，暂时只禁用 vapor 模式
                if (isDom2) {
                    const plugin = config.plugins.find((p) => p.name === 'vite:vue');
                    if (plugin?.api?.options) {
                        plugin.api.options.devToolsEnabled = false;
                        plugin.api.options.isProduction = true;
                        // TODO 临时禁用，目前有bug 等待 https://github.com/vuejs/core/pull/13630 合并
                        // 使用内部自己定义的 transformAssetUrls
                        plugin.api.options.template.transformAssetUrls = false;
                    }
                }
                if (isDom2 && isHarmony) {
                    // 仅鸿蒙 dom2 开启：该模式会拆出大量 ESM chunk，
                    // 增量编译时未变更 chunk 也会反复走 Vite 的 esbuild renderChunk。
                    // 其他平台/模式先不改，避免扩大影响面。
                    initHarmonyDom2EsbuildTranspileCache(config);
                }
            },
            generateBundle(_, bundle) {
                // 调整所有sourceMap文件
                const currentSourceMapFiles = enableSourceMapIncremental
                    ? new Set()
                    : undefined;
                Object.entries(bundle).forEach(([file, asset]) => {
                    if (file.endsWith('.js.map') && asset.type === 'asset') {
                        currentSourceMapFiles?.add(file);
                        const sourceMapHash = enableSourceMapIncremental
                            ? (0, uni_cli_shared_1.hash)(asset.source)
                            : '';
                        // 鸿蒙 DOM2 开发模式下，内容未变化的 sourcemap 无需重复 parse 和写入。
                        if (enableSourceMapIncremental &&
                            sourceMapHashCache.get(file) === sourceMapHash) {
                            return;
                        }
                        (0, sourceMap_1.writeAppServiceSourceMapToCache)({
                            file,
                            sourceMap: asset.source,
                            bundle,
                            inputDir,
                            outputDir,
                            cacheDir: process.env.UNI_APP_X_CACHE_DIR,
                            keepSourceMapInBundle: process.env.UNI_PLATFORM === 'app-harmony',
                            useCacheSourceMapUrl: process.env.NODE_ENV === 'development' &&
                                ((isAndroid && isDom2) || isIOS),
                            sourceMapUrlMode: isAndroid ? 'absolute' : 'relative',
                            sourceRootMode: isIOS ? 'absolute' : 'relative',
                        });
                        if (enableSourceMapIncremental) {
                            sourceMapHashCache.set(file, sourceMapHash);
                        }
                    }
                });
                if (currentSourceMapFiles) {
                    sourceMapHashCache.forEach((_, file) => {
                        if (!currentSourceMapFiles.has(file)) {
                            sourceMapHashCache.delete(file);
                        }
                    });
                }
            },
            async writeBundle() {
                // x 上暂时编译所有uni ext api，不管代码里是否调用了
                // 框架内部编译时，不需要
                if (process.env.UNI_COMPILE_TARGET !== 'ext-api') {
                    await (0, uni_cli_shared_1.buildUniExtApis)(process.env.UNI_UTS_PLATFORM === 'app-harmony');
                    await (0, uni_cli_shared_1.buildNonTreeShakingUniModules)();
                }
            },
        };
    };
}
exports.createUniAppJsEnginePlugin = createUniAppJsEnginePlugin;
